aea/cli/local_registry_sync.py

```diff
@@ -0,0 +1,140 @@
+#
+#   Copyright 2018-2019 Fetch.AI Limited
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+"""Implementation of the 'aea sync-local-registry' subcommand."""
+import os
+import shutil
+from pathlib import Path
+from tempfile import TemporaryDirectory
+from typing import Generator, Tuple, Union
+
+import click
+
+from aea.cli.fingerprint import determine_package_type_for_directory
+from aea.cli.registry.add import fetch_package
+from aea.cli.registry.utils import get_package_meta
+from aea.cli.utils.context import Context
+from aea.cli.utils.decorators import pass_ctx
+from aea.configurations.data_types import PackageId, PackageType, PublicId
+from aea.configurations.loader import load_component_configuration
+
+
+PACKAGES_DIRS = [i.to_plural() for i in PackageType]
+
+
+@click.command()
+@pass_ctx
+def local_registry_sync(ctx: Context):
+    """Upgrade the local package registry."""
+    do_local_registry_update(ctx.cwd)
+
+
+def do_local_registry_update(base_dir: Union[str, Path]) -> None:
+    """
+    Perform local registry update.
+
+    :param base_dir: root directory of the local registry.
+
+    :return: None
+    """
+    for package_id, package_dir in enlist_packages(base_dir):
+        current_public_id = package_id.public_id
+        latest_public_id = get_package_latest_public_id(package_id)
+        if not (  # pylint: disable=superfluous-parens
+            current_public_id < latest_public_id
+        ):
+            click.echo(f"{package_id} is the latest version.")
+            continue
+        click.echo(f"Updating to {latest_public_id}.")
+        replace_package(str(package_id.package_type), latest_public_id, package_dir)
+
+
+def replace_package(
+    package_type: str, public_id: PublicId, package_dir: Union[Path, str]
+) -> None:
+    """
+    Download, extract and replace exists package.
+
+    :param package_type: str.
+    :param public_id: pacakge bulic id to download
+    :param: package_dir: target package dir
+
+    :return: None
+    """
+    with TemporaryDirectory() as tmp_dir:
+        new_package_dir = os.path.join(tmp_dir, public_id.name)
+        os.mkdir(new_package_dir)
+        fetch_package(
+            package_type, public_id=public_id, cwd=tmp_dir, dest=new_package_dir
+        )
+        shutil.rmtree(package_dir)
+        shutil.move(new_package_dir, package_dir)
```

Comment:
If there is a failure here, we should handle it and keep the old package.

Reply:
it does this logic fetch first, remove then